loudml/opentsdb.py

```diff
@@ -0,0 +1,364 @@
+                self.url,
+                time_pred,
+                AGGREGATORS.get(q["aggregator"]),
+                q["down_sampler"],
+                q["metric_name"],
+                str(q["tags"])
+            )
+            # TODO: OpenTSDB is capable of running multiple subquries in
+            # one shot. Refactor it to use one request to server
+            resp = requests.get(query_url)
+
+            if resp.status_code >= 200 and resp.status_code < 400:
+                results.append(OpenTSDBResult(resp.json()))
+            elif "No such name for" in resp.text:
+                # Known OpenTSDB issue with 400 and trace if metric not found
+                # https://github.com/OpenTSDB/opentsdb/issues/792
+                results.append(OpenTSDBResult([]))
+            else:
+                logging.error('OpenTSDB error',
+                    resp.status_code, resp.text[:200])
+
+        return results
+
+    def put(self, entry):
+        """
+        Store a point in database
+        """
+        url = "%s/api/put" % self.url
+
+        if entry["sync"]:
+            url = "{}?sync".format(url)
+
+        del entry["sync"]
+
+        resp = requests.post(url, json=entry) # data=entry ?
+
+        if resp.status_code >= 200 and resp.status_code < 400:
+            pass # everything is ok, should we indicate it somehow?
+        else:
+            logging.error('OpenTSDB error', resp.status_code, resp.text[:200])
+
+
+class OpenTSDBBucket(Bucket):
+    """
+    OpenTSDB bucket
+    """
+
+    SCHEMA = Bucket.SCHEMA.extend({
+        Required('addr'): str,
+        Optional('use_ssl', default=False): Boolean(),
+    })
+
+    def __init__(self, cfg):
+        cfg['type'] = 'opentsdb'
+        super().__init__(cfg)
+        self._opentsdb = None
+
+    @property
+    def addr(self):
+        return self.cfg['addr']
+
+    @property
+    def use_ssl(self):
+        return self.cfg.get('use_ssl') or False
+
+    @property
+    def opentsdb(self):
+        if self._opentsdb is None:
+            addr = parse_addr(self.addr, default_port=4242)
+            logging.info(
+                "connecting to opentsdb on %s:%d",
+                addr['host'],
+                addr['port'],
+            )
+            self._opentsdb = OpenTSDBClient(
+                host=addr['host'],
+                port=addr['port'],
+                ssl=self.use_ssl,
+            )
+
+        return self._opentsdb
+
+    def insert_data(self, data):
+        raise errors.NotImplemented("OpenTSDB is a pure time-series database")
+
+    def insert_times_data(
+        self,
+        ts,
+        data,
+        measurement=None,
+        tags={},
+        sync=False,
+        *args,
+        **kwargs
+    ):
+        """
+        Insert data
+
+        'sync' - Whether or not to wait for the data to be flushed to storage
+        before returning the results. Used in tests
+        """
+        ts = int(make_ts(ts))
+        for k, v in data.items():
```

Comment:
slightly more Pythonic: 

filtered = filter(lambda item: item[1] is not None, data.items())
map
for k, v in filtered: